raft-tech/TANF-app
# feat(1170): Enforce AMS usage for ACF Staff

### Summary of Changes
_Provide a brief summary of changes_
Pull request closes #1170

### How to Test
While this functionality can be tested locally to some extent, to ensure AMS functionality of #1136 remains intact, this must also be tested in the tanf-dev raft space so please ensure this branch is deployed there.

Testing locally:
1. Spin up the environment locally using
```
cd tdrs-frontend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d
cd ../tdrs-backend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
cd ../ ; docker ps | grep tdp
docker logs -f <container id from previous command>
```
2. Navigate to the frontend at `localhost:3000` and click "Sign in with Login.gov for grantees"
3. Sign in with your `@acf.hhs.gov` email and password
4. Ensure you are not able to sign in -- the backend should reply with a 403 Forbidden Error.
5. You may proceed with regular reassurance tests such as using login.gov with a non-ACF e-mail
6. Due to this being a local environment, it is expected that ACF AMS authentication will always fail.

Testing via tanf-dev raft:
1. Navigate to `https://tdp-frontend-raft.app.cloud.gov` and click "Sign in with Login.gov for grantees"
1. Proceed with your `@acf.hhs.gov` e-mail user using Login.gov.
1. Ensure you are not able to sign in -- the backend should reply with a 403 Forbidden Error.
2. Sign out and now click "Sign in with ACF AMS for ACF Staff".
3. Proceed with your PIV/CAC card and ensure you can login successfully.

> *Demo GIF(s) and screenshots for testing procedure*
![ezgif com-gif-maker](https://user-images.githubusercontent.com/84722778/149580557-88e8b28f-a585-4df0-bff6-a7b725cbe748.gif)

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

 - [x] ACF staff user is not actually authenticated by Login.gov (not able to browse/utilize TDP)
 - [x] If an ACF staff user uses  login.gov for authentication,  an error is returned indicating incorrect authentication method
 - [x] First-time ACF user story is handled for a user not yet assigned a role
 - [x] Ensure logout is working as expected
 _mostly works but is buggy as noted [here](https://github.com/raft-tech/TANF-app/pull/1536#issuecomment-1017937447)
 - [x] @ADPennington reviewed and tested

As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*


### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/Technical-Documentation/how-government-will-test-a11y.md)

_n/a_

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful deployment by assigning a 'Deploy with CircleCI - {env}' label*


### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*


### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

_no new issues detected_

### Deliverable 8: Context

> * Performance Standard(s): Code must be understandable and contextualized for the reviewers possess the knowledge and background necessary for analysis and constructive criticism to take place.

> * Acceptable Quality Level: Code submitted in the pull request has context.*
